File: modyn/storage/internal/filesystem_wrapper/local_filesystem_wrapper.py
"""Local filesystem wrapper.

This module contains the local filesystem wrapper.
It is used to access files on the local filesystem.
"""
import os

from modyn.storage.internal.filesystem_wrapper.abstract_filesystem_wrapper import AbstractFileSystemWrapper
from modyn.storage.internal.filesystem_wrapper.filesystem_wrapper_type import FilesystemWrapperType


class LocalFilesystemWrapper(AbstractFileSystemWrapper):
    """Local filesystem wrapper."""

    def __init__(self, base_path: str):
        """Init local filesystem wrapper.

        Args:
            base_path (str): Base path of local filesystem
        """
        super().__init__(base_path)
        self.filesystem_wrapper_type = FilesystemWrapperType.LocalFilesystemWrapper

    def __is_valid_path(self, path: str) -> bool:
        return path.startswith(self.base_path)

    def _get(self, path: str) -> bytes:
        """Get file content.

        Args:
            path (str): Absolute path to file

        Raises:
            FileNotFoundError: If path is not valid
            IsADirectoryError: If path is a directory

        Returns:
            bytes: File content
        """
        if not self.__is_valid_path(path):
            raise ValueError(f"Path {path} is not valid.")
        if not self.isfile(path):
            raise IsADirectoryError(f"Path {path} is a directory.")
        with open(path, "rb") as file:
            return file.read()

    def exists(self, path: str) -> bool:
        """Check if path exists.

        Args:
            path (str): Absolute path to file or directory

        Returns:
            bool: True if path exists, False otherwise
        """
        return os.path.exists(path)

    def list(self, path: str, recursive: bool = False) -> list[str]:
        """List files in directory.

        Args:
            path (str): Absolute path to directory
            recursive (bool, optional): List files recursively. Defaults to False.

        Raises:
            ValueError: If path is not valid
            NotADirectoryError: If path is not a directory

        Returns:
            list[str]: List of files in directory
        """
        if not self.__is_valid_path(path):
            raise ValueError(f"Path {path} is not valid.")
        if not self.isdir(path):
            raise NotADirectoryError(f"Path {path} is not a directory.")
        if recursive:
            return [os.path.join(dp, f) for dp, dn, fn in os.walk(os.path.expanduser(path)) for f in fn]
        return os.listdir(path)

    def isdir(self, path: str) -> bool:
        """Check if path is a directory.

        Args:
            path (str): Absolute path to directory

        Returns:
            bool: True if path is a directory, False otherwise
        """
        return os.path.isdir(path)

    def isfile(self, path: str) -> bool:
        """Check if path is a file.

        Args:
            path (str): Absolute path to file

        Returns:
            bool: True if path is a file, False otherwise
        """
        return os.path.isfile(path)

    def get_size(self, path: str) -> int:
        """Get size of file.

        Args:
            path (str): Absolute path to file

        Raises:
            ValueError: If path is not valid
            IsADirectoryError: If path is a directory

        Returns:
            int: Size of file in bytes
        """
        if not self.__is_valid_path(path):
            raise ValueError(f"Path {path} is not valid.")
        if not self.isfile(path):
            raise IsADirectoryError(f"Path {path} is a directory.")
        return os.path.getsize(path)

    def get_modified(self, path: str) -> int:
        """Get modification time of file.

        Args:
            path (str): Absolute path to file

        Raises:
            ValueError: If path is not valid
            IsADirectoryError: If path is a directory

        Returns:
            int: Modification time in milliseconds rounded to the nearest integer
        """
        if not self.__is_valid_path(path):
            raise ValueError(f"Path {path} is not valid.")
        if not self.isfile(path):
            raise IsADirectoryError(f"Path {path} is a directory.")
        return int(os.path.getmtime(path) * 1000)

    def get_created(self, path: str) -> int:
        """Get creation time of file.

        Args:
            path (str): Absolute path to file

        Raises:
            ValueError: If path is not valid
            IsADirectoryError: If path is a directory

        Returns:
            int: Creation time in milliseconds rounded to the nearest integer
        """
        if not self.__is_valid_path(path):
            raise ValueError(f"Path {path} is not valid.")
        if not self.isfile(path):
            raise IsADirectoryError(f"Path {path} is a directory.")
        return int(os.path.getctime(path) * 1000)

    def join(self, *paths: str) -> str:
        """Join paths.

        Returns:
            str: Joined path
        """
        return os.path.join(*paths)

    def delete(self, path: str) -> None:
        """Delete file.

        Args:
            path (str): Absolute path to file

        Raises:
            ValueError: If path is not valid
            IsADirectoryError: If path is a directory
        """
        return os.remove(path)

File: modyn/storage/src/internal/filesystem_wrapper/local_filesystem_wrapper.cpp
#include "internal/filesystem_wrapper/local_filesystem_wrapper.hpp"

#include <fmt/format.h>
#include <spdlog/spdlog.h>
#include <sys/stat.h>
#include <sys/types.h>

#include <chrono>
#include <filesystem>
#include <fstream>
#include <iostream>
#include <stdexcept>
#include <vector>

#include "modyn/utils/utils.hpp"

using namespace modyn::storage;

std::vector<unsigned char> LocalFilesystemWrapper::get(const std::string& path) {
  std::ifstream file;
  file.open(path, std::ios::binary);
  std::vector<unsigned char> buffer(std::istreambuf_iterator<char>(file), {});
  file.close();
  return buffer;
}

std::shared_ptr<std::ifstream> LocalFilesystemWrapper::get_stream(const std::string& path) {
  std::shared_ptr<std::ifstream> file = std::make_shared<std::ifstream>();
  file->open(path, std::ios::binary);
  return file;
}

bool LocalFilesystemWrapper::exists(const std::string& path) { return std::filesystem::exists(path); }

std::vector<std::string> LocalFilesystemWrapper::list(const std::string& path, bool recursive, std::string extension) {
  std::vector<std::string> paths;

  if (!std::filesystem::exists(path)) {
    return paths;
  }

  if (recursive) {
    for (const auto& entry : std::filesystem::recursive_directory_iterator(path)) {
      const std::filesystem::path& entry_path = entry.path();
      if (!std::filesystem::is_directory(entry) && entry_path.extension().string() == extension) {
        paths.push_back(entry_path);
      }
    }
  } else {
    for (const auto& entry : std::filesystem::directory_iterator(path)) {
      const std::filesystem::path& entry_path = entry.path();
      if (!std::filesystem::is_directory(entry) && entry_path.extension().string() == extension) {
        paths.push_back(entry_path);
      }
    }
  }

  return paths;
}

bool LocalFilesystemWrapper::is_directory(const std::string& path) { return std::filesystem::is_directory(path); }

bool LocalFilesystemWrapper::is_file(const std::string& path) { return std::filesystem::is_regular_file(path); }

uint64_t LocalFilesystemWrapper::get_file_size(const std::string& path) {
  return static_cast<int64_t>(std::filesystem::file_size(path));
}

int64_t LocalFilesystemWrapper::get_modified_time(const std::string& path) {
  ASSERT(is_valid_path(path), fmt::format("Invalid path: {}", path));
  ASSERT(exists(path), fmt::format("Path does not exist: {}", path));
  static_assert(sizeof(int64_t) >= sizeof(std::time_t), "Cannot cast time_t to int64_t");

  // there is no portable way to get the modified time of a file in C++17 and earlier
  // clang-format off
  struct stat file_attribute {};//This line keeps getting changed by clang format in my machine and not passing the format test when pushed.
  // clang-format on
  stat(path.c_str(), &file_attribute);
  return static_cast<int64_t>(file_attribute.st_mtime);
  /* C++20 version, not supported by compilers yet */
  /*
    const auto modified_time = std::filesystem::last_write_time(path);
    const auto system_time = std::chrono::clock_cast<std::chrono::system_clock>(modified_time);
    const std::time_t time = std::chrono::system_clock::to_time_t(system_time);
    return static_cast<int64_t>(time); */
}

bool LocalFilesystemWrapper::is_valid_path(const std::string& path) { return std::filesystem::exists(path); }

bool LocalFilesystemWrapper::remove(const std::string& path) {
  ASSERT(!std::filesystem::is_directory(path), fmt::format("Path is a directory: {}", path));

  if (!std::filesystem::exists(path)) {
    SPDLOG_WARN(fmt::format("Trying to delete already deleted file {}", path));
    return true;
  }

  SPDLOG_DEBUG("Removing file: {}", path);

  return std::filesystem::remove(path);
}

FilesystemWrapperType LocalFilesystemWrapper::get_type() { return FilesystemWrapperType::LOCAL; }
